python-packages/presto_drf/presto_drf/mixins.py

```diff
@@ -0,0 +1,77 @@
+from typing import List
+
+
+class SerializerOptInFieldsMixin:
+    """
+    Regulates fields exposed by default & as requested based on query parameters or context.
+
+    Pass 'include_fields' / 'opt_in_fields' thru query params or context to use; multiple fields are separated by comma eg,
+    /?include_fields=first_name,email&opt_in_fields=gait_recognition_prediction
+
+    1. By default, all "fields" defined in serializer, minus those listed in "opt_in_fields" would be returned.
+    2. If "include_fields" is supplied, only fields requested this way would be returned.
+    3. If "opt_in_fields" is supplied, fields requested this way PLUS fields from #1 or #2 would be returned.
+
+    Pinned fields are always returned. (currently pk only if pk exists)
+    """
+
+    def get_pinned_fields(self) -> List:
+        """
+        Get by-default pinned fields. Pinned fields are fields always returned regardless of include_fields inclusions.
+        Override on serializer to customize.
+
+        Currently, the only pinned field is pk of model (with ModelSerializer).
+
+        :return: [] - a list of pinned fields.
+        """
+
+        try:
+            return [self.Meta.model._meta.pk.name]
+        except Exception:
+            pass
+        return []
+
+    def __init__(self, *args, **kwargs):
+        result = super().__init__(*args, **kwargs)
+
+        if not hasattr(
+            self, "context"
+        ):  # serializer invoked without context - inspection?
+            return result
+
+        pinned_fields = self.get_pinned_fields()
```

Comment:
Here is a simplified version (at least I see it as simpler, esp re handling of things that may (not) be present, eg query_params)

```py
try:
	query_params = self.context['request'].query_params
except KeyError, AttributeError:
	query_params = {}

def parse_includes(spec: Union[str, Iterable[str]]) -> Set[str]:
    # turns query params with a source of 
    #  ?include_fields=first_name&include_fields=last_name,email
    # into ['first_name', 'last_name', 'email']
    if isinstance(spec, str): spec = [spec]
	return set(flatten(f.split(',') for f in spec)))

try:
	include_fields = parse_includes(query_params['include_fields'])
except KeyError:
	include_fields = set(self.Meta.fields) - getattr(self.Meta, "opt_in_only_fields", set())

try:
	opt_in_fields = parse_includes(query_params['opt_in_fields'])
except KeyError:
	opt_in_fields = set()

final_fields = set(self.get_pinned_fields()) + include_fields + opt_in_fields

# this ignores misspelled fields; can iterate through final_fields to build up self.fields 
# if you want to throw an error instead
self.fields = {
	name: field
	for name, field
	in self.fields
	if name in final_fields
}
```

(Haven't tested this, in particular `parse_includes` may be wrong: my recollection of the behaviour of django's query parser is:
-  `?a=x` => `query_params['a'] == 'x'`
- `?a=x&a=y` => `query_params['a'] == ['x', 'y']`